### PR TITLE
Android 11 inline autofill

### PR DESCRIPTION
**Pull request by mportune-bw**
Support for Android 11 inline autofill. If a fill request includes an `InlineSuggestionsRequest` (requires Android 11 plus an IME (keyboard) that supports inline presentations) then we construct the additional views according to the given restrictions (dictated by the IME within the request) and returned along with the original overlay views.  The result is a horizontally-scrollable collection of ciphers anchored within the IME.  The original overlay is used if no `InlineSuggestionsRequest` is supplied, or if anything goes wrong during the creation of the inline presentation.

Some tidbits:

- With the exception of our logo, the inline icons are tinted according to the rules in the `InlineSuggestionsRequest` to allow the IME to drive the theme.
- With some browsers this can be finicky if the browser's own autofill functions are still enabled (varies, of course)
- Firefox (prod & beta) is currently broken as it keeps re-initializing the keyboard and erasing the inline presentation.  This has been fixed in their nightly builds.

Screen Shot 2020-11-10 at 11 39 30 AM

### CODE REVIEW

File: src/Android/Autofill/AutofillHelpers.cs [hunk 1]
```diff
@@ -132,39 +142,102 @@ public static async Task<List<FilledItem>> GetFillItemsAsync(Parser parser, ICip
...
+                    var inlineSuggestionsRequest = fillRequest.InlineSuggestionsRequest;
+                    if (inlineSuggestionsRequest != null)
+                    {
+                        inlineMaxSuggestedCount = inlineSuggestionsRequest.MaxSuggestionCount;
+                    }
+                    inlinePresentationSpecs = inlineSuggestionsRequest?.InlinePresentationSpecs;
+                    if (inlinePresentationSpecs != null)
+                    {
+                        inlinePresentationSpecsCount = inlinePresentationSpecs.Count;
+                    }
```
Comment (cscharf) [inline src/Android/Autofill/AutofillHelpers.cs:165]: This block can be simplified:
```c#
inlineMaxSuggestedCount = fillRequest.InlineSuggestionsRequest?.MaxSuggestionCount ?? 0;
inlinePresentationSpecsCount = inlineSuggestionsRequest?.InlinePresentationSpecs?.Count ?? 0;
```

**Review by cscharf — COMMENTED** [review]
Just minor feedback on simplifying one of the blocks, but otherwise looks good!

File: src/Android/Autofill/AutofillHelpers.cs [hunk 2]
```diff
@@ -132,39 +142,102 @@ public static async Task<List<FilledItem>> GetFillItemsAsync(Parser parser, ICip
...
+        public static FillResponse BuildFillResponse(Parser parser, List<FilledItem> items, bool locked,
+            FillRequest fillRequest = null)
         {
+            // Acquire inline presentation specs on Android 11+
+            IList<InlinePresentationSpec> inlinePresentationSpecs = null;
+            var inlinePresentationSpecsCount = 0;
+            var inlineMaxSuggestedCount = 0;
+            if (fillRequest != null)
+            {
+                if ((int)Build.VERSION.SdkInt >= 30)
```
Comment (kspearrin) [inline src/Android/Autofill/AutofillHelpers.cs:154]: Can also combine these conditions together in one `if`

**Review by kspearrin — APPROVED** [review]

**Comment by yurividal** [comment]
Should this also work on Samsung's Android 11?

**Comment by mportune-bw** [comment]
> Should this also work on Samsung's Android 11?

There's no reason it _shouldn't_, however in the end that's up to Samsung.

**Comment by yurividal** [comment]
I just verified that it indeed works! Although, only on Chrome, and using the Samsung keyboard stock keyboard.
Any attempt to use Samsung Internet, Samsung internet beta, and/or gboard does not work.
I guess now we just need to wait for Samsung to implement it on their browser, and that other keyboards such as SwiftKey implement it too. Thank you bitwarden team!!

**Comment by yurividal** [comment]
> > Should this also work on Samsung's Android 11?
> 
> There's no reason it _shouldn't_, however in the end that's up to Samsung.

I just verified that it indeed works! Although, only on Chrome, and using the Samsung keyboard stock keyboard.
Any attempt to use Samsung Internet, Samsung internet beta, and/or gboard does not work.
I guess now we just need to wait for Samsung to implement it on their browser, and that other keyboards such as SwiftKey implement it too. Thank you bitwarden team!!

**Comment by mportune-bw** [comment]
For anyone using Brave, it seems to be working when all three of Brave's autofill features are disabled (passwords, payment methods, and Addresses).  Just FYI.

**Comment by wallismith** [comment]
@mportune-bw Unfortunately, that doesn't fix the issue for Brave. It doesn't seem to work 90% of the time. I'm almost certain it's not code related. This has been an ongoing issue with Chromium browsers on Android, even before the new Android 11 keyboard auto fill API.

Looks like someone submitted a patches for this issue: https://github.com/bitwarden/mobile/issues/489#issuecomment-725918299

**Comment by mportune-bw** [comment]
@wallismith Sorry to hear that.  Autofill can be incredibly frustrating since there are so many moving parts outside of our control.  Just curious what device are you testing with?  (I'm on a Pixel 4a)

**Comment by wallismith** [comment]
@mportune-bw I'm on a Pixel 3 using Gboard.

**Comment by bingoxo** [comment]
Does this work with OpenBoard ,FlorisBoard ,AnySoftKeyboard or any other FOSS keyboard ?

**Comment by shivramg94** [comment]
@bingoxo I could only see it working on Gboard and Samsung keyboard.

**Pull request merged**